### PR TITLE
Autocli enhancement - using FileDescriptorSet to describe input and output types

### PR DESCRIPTION
**Pull request by tomaszrybicki**

### CODE REVIEW

File: source/octf/cli/internal/cmd/Command.h [hunk 1]
```diff
@@ -24,6 +24,7 @@ class Command : public ICommand {
      * @brief Empty command constructor
      */
     Command();
+    // TODO: implement setting cmd ops
```
Comment (tomaszrybicki) [inline source/octf/cli/internal/cmd/Command.h]: Remove

File: source/octf/cli/internal/cmd/CommandProtobuf.cpp [hunk 2]
```diff
@@ -57,52 +57,36 @@ CommandProtobuf::CommandProtobuf(const proto::CliCommand &cmdDesc)
     // Set name of file descriptor
     m_fd.set_name("cli");
```
Comment (tomaszrybicki) [inline source/octf/cli/internal/cmd/CommandProtobuf.cpp]: Remove this unused member, and possibly others

File: source/octf/cli/internal/cmd/CommandProtobuf.cpp [hunk 3]
```diff
@@ -57,52 +57,36 @@ CommandProtobuf::CommandProtobuf(const proto::CliCommand &cmdDesc)
...
+                " to build input message description");
     }
 
-    m_outDesc = m_descPool.FindMessageTypeByName("output");
+    m_outDesc =
+            m_descPool.FindMessageTypeByName(cmdDesc.outputmessagetypename());
     if (!m_outDesc) {
         throw InvalidParameterException(
                 "Cannot build output parameter description.");
     }
```
Comment (tomaszrybicki) [inline source/octf/cli/internal/cmd/CommandProtobuf.cpp]: Unify Exception messages

File: source/octf/interface/InterfaceCliImpl.h [hunk 4]
```diff
@@ -9,6 +9,8 @@
 #include <google/protobuf/descriptor.h>
 #include <google/protobuf/service.h>
 #include <octf/node/INode.h>
+// TODO
+// #include "proto/InterfaceCLI.pb.h"
```
Comment (tomaszrybicki) [inline source/octf/interface/InterfaceCliImpl.h]: Remove

File: source/octf/interface/InterfaceCliImpl.h [hunk 5]
```diff
@@ -23,25 +25,26 @@ class InterfaceCliImpl : public proto::InterfaceCli {
...
+    //     * @param param Parameter to be defined
+    //     * @param fieldDesc Field descriptor with source data
+    //     * @param isInput Is this parameter used for input, if yes it is
+    //     validated
+    //     * @return Result of operation
+    //     */
+    //    static bool setParamDescription(
+    //            proto::CliParameter *param,
+    //            const google::protobuf::FieldDescriptor *fieldDesc,
+    //            bool isInput = true);
```
Comment (tomaszrybicki) [inline source/octf/interface/InterfaceCliImpl.h]: Remove

File: source/octf/interface/InterfaceCliImpl.h [hunk 6]
```diff
@@ -61,22 +64,23 @@ class InterfaceCliImpl : public proto::InterfaceCli {
...
-    static void setEnumOps(const google::protobuf::EnumDescriptor *enumDesc,
-                           proto::CliParameter *param);
+    //    static void setEnumOps(const google::protobuf::EnumDescriptor
+    //    *enumDesc,
+    //                           proto::CliParameter *param);
 
-    static proto::CliParameter_Type getTypeFromCppType(
-            google::protobuf::FieldDescriptor::CppType cppType);
+    //    static proto::CliParameter_Type getTypeFromCppType(
+    //            google::protobuf::FieldDescriptor::CppType cppType);
```
Comment (tomaszrybicki) [inline source/octf/interface/InterfaceCliImpl.h]: Remove

File: source/octf/proto/InterfaceCLI.proto [hunk 7]
```diff
@@ -9,38 +9,46 @@ import "defs.proto";
...
+//
+//    // Parameter options
+//    OptsParam paramOps = 3;
+//
+//    // If type is message, this field describes message's fields
+//    repeated CliParameter nestedParam = 4;
+//
+//    // Can this parameter hold multiple values
+//    bool isMultipleValue = 5;
+//}
```
Comment (tomaszrybicki) [inline source/octf/proto/InterfaceCLI.proto]: Remove

File: source/octf/utils/ProtoConverter.h [hunk 8]
```diff
@@ -6,15 +6,51 @@
...
+ * @brief This class allows to create a FileDescriptorSet object, which holds
+ * the definition of all supplied messages
+ */
+class FileDescriptorSetCreator {
+public:
+    FileDescriptorSetCreator() = default;
+    virtual ~FileDescriptorSetCreator() = default;
+
+    /**
+     * @brief Create a FileDescriptorSet from added message's descriptions
```
Comment (tomaszrybicki) [inline source/octf/utils/ProtoConverter.h]: messages'

File: source/octf/utils/ProtoConverter.h [hunk 9]
```diff
@@ -6,15 +6,51 @@
...
+     * @return FileDescriptorSet object
+     */
+    google::protobuf::FileDescriptorSet createFileDescriptorSet();
+
+    /**
+     * @brief Add a message description to FileDescriptorSet.
+     *
+     * Message from package "foo.bar" will be put in a FileDescriptor of name
+     * "foo.bar". Any message's field which are of type message or enum, will be
```
Comment (tomaszrybicki) [inline source/octf/utils/ProtoConverter.h]: fields

File: source/octf/proto/InterfaceCLI.proto [hunk 10]
```diff
@@ -52,31 +30,27 @@ message CliCommand {
...
 
-    // Array of CLI command parameters
-    repeated CliParameter inputParameter = 5;
+    // Descriptors of input and output types for this command
+    google.protobuf.FileDescriptorSet inputOutputDescription = 5;
 
-    // Array of CLI parameters defining method response
-    repeated CliParameter outputParameter = 6;
-}
+    // Type name of input message for this command
```
Comment (mariuszbarczak) [inline source/octf/proto/InterfaceCLI.proto]: I believe it's full name, so let's emphasize it in comment

File: source/octf/utils/ProtoConverter.cpp [hunk 11]
```diff
@@ -96,5 +100,133 @@ int decodeVarint32(const uint8_t *buffer, uint64_t size, int &value) {
...
+        const FieldDescriptor *field = msgDesc->field(i);
+
+        // Handle message type
+        if (field->type() == FieldDescriptor::Type::TYPE_MESSAGE) {
+            // Verify we only describe each type once
+            if (std::find(m_knownTypes.begin(), m_knownTypes.end(),
+                          field->full_name()) == m_knownTypes.end()) {
+                // If we use a message type defined in different package, we
+                // need to add appropriate import to FileDescriptor.
+                std::string newTypePackage =
```
Comment (mariuszbarczak) [inline source/octf/utils/ProtoConverter.cpp]: const auto&

File: source/octf/utils/ProtoConverter.cpp [hunk 12]
```diff
@@ -96,5 +100,133 @@ int decodeVarint32(const uint8_t *buffer, uint64_t size, int &value) {
...
+                if (fd.package() != newTypePackage) {
+                    // If not already imported
+                    if (!contains<std::string>(fd.dependency(),
+                                               newTypePackage)) {
+                        fd.add_dependency(newTypePackage);
+                    }
+                }
+
+                // Define enum
+                std::string package = field->enum_type()->file()->package();
```
Comment (mariuszbarczak) [inline source/octf/utils/ProtoConverter.cpp]: the same like in line 178

File: source/octf/utils/ProtoConverter.cpp [hunk 13]
```diff
@@ -96,5 +100,133 @@ int decodeVarint32(const uint8_t *buffer, uint64_t size, int &value) {
...
+                EnumDescriptorProto *enumDesc =
+                        getFileDesc(package).add_enum_type();
```
Comment (mariuszbarczak) [inline source/octf/utils/ProtoConverter.cpp]: use local var for: getFileDesc(package)

File: source/octf/utils/ProtoConverter.cpp [hunk 14]
```diff
@@ -96,5 +100,133 @@ int decodeVarint32(const uint8_t *buffer, uint64_t size, int &value) {
...
+                }
+
+                // Recursively add field's (of type message) definitions
+                addMessageDesc(field->message_type());
+                m_knownTypes.push_back(field->full_name());
+            }
+        }
+
+        // Handle enum type
+        if (field->type() == FieldDescriptor::Type::TYPE_ENUM) {
```
Comment (mariuszbarczak) [inline source/octf/utils/ProtoConverter.cpp]: else if

File: source/octf/utils/ProtoConverter.cpp [hunk 15]
```diff
@@ -101,8 +101,9 @@ int decodeVarint32(const uint8_t *buffer, uint64_t size, int &value) {
 }
 
 template <class ElementType>
-bool contains(const google::protobuf::RepeatedPtrField<ElementType> array,
-              ElementType value) {
+bool doesRepeatedPtrFieldContain(
+        const google::protobuf::RepeatedPtrField<ElementType> array,
+        ElementType value) {
```
Comment (mariuszbarczak) [inline source/octf/utils/ProtoConverter.cpp]: const ElementType& value

**Comment by mariuszbarczak** [comment]
Fixes #50

**Pull request merged**